rfd/0078-role-variables.md

```diff
@@ -0,0 +1,264 @@
+      # form the complete list of values for the variable.
+      values:
+        # The simplest value is a list of static strings.
+        - out: [ubuntu]
+
+        # Values can also contain traits and transforms.
+        - out: ['{{regexp.replace(external.username, "-", "_")}}']
+
+        # Values can also take an input, which will usually be a trait.
+        - input: "{{external.email}}"
+          # match is a regular expression which must match the input in order
+          # for this value to be included.
+          match: "^nic@goteleport.com$"
+          out: [root]
+
+        - input: "{{external.email}}"
+          # match may contain contain capture groups which can be expanded in
+          # the output.
+          #
+          # Here "alice@goteleport.com" maps to "alice", but "foo@example.com"
+          # will be rejected.
+          match: "^(.*)@goteleport.com$"
+          out: ["$1"]
+
+    - name: "allow-env"
+      values:
+        # An input will expand to a list of N inputs if the trait is a list of size N.
+        # Any input value which matches the `match` regex will contribute one
+        # output value, which may contain regex captures from that specific input.
+        #
+        # Here [devs, env-staging, env-prod] maps to [staging, prod].
+        - input: "{{external.groups}}"
+          match: '^env-(\w+)$'
+          out: ['$1']
+
+        # If any of the N input values match the `not_match` regex, none of the
+        # values will be used.
+        #
+        # Here [devs] maps to [dev], but [devs, contractors] maps to []
+        - input: "{{external.groups}}"
+          match: '^devs$'
+          not_match: 'contractors'
+          out: ["dev"]
+  allow:
+    logins:
+      # Variables can be referenced as `vars.<role_name>.<variable_name>` or
+      # `vars["<role_name>"]["<variable_name"]`.
+      - "{{vars.node_users.logins}}"
+    node_labels:
+      env: ["{{vars.node_users["allow-env"]}}"]
+```
+
+These variables could be used in another role without redefinition:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: app-users
+spec:
+  allow:
+    app_labels:
+      env: ["{{vars.node_users["allow-env"]}}"]
+```
+
+### How to reference variables
+
+Variables can be referenced similar to traits, by enclosing their name in
+`{{}}`.
+
+Variable names are prefixed by that `vars` namespace and include the role name
+to avoid collisions.
+
+You can use the selector syntax `{{vars.role_name.variable_name}}`, the index
+syntax `{{vars["role-name"]["variable-name"]}}`, or some combination
+`{{vars[role_name].variable_name}}`.
+
+It is necessary to use the index syntax and enclose the identifier (role or
+variable name) in quotes if it is not a valid
+[Go identifier](https://go.dev/ref/spec#Identifiers),
+or else the parser will not be able to handle it.
+The same is true for traits today.
+
+Variables can be referenced in the role which defines them, or in any other
```

Comment:
Let's keep variables role-local. This adds some inconveniences, but has major upside in simplicity of evaluation and the lack of ambiguity. Linter can validate any role just by looking at it, and having no other context.